feat(Product) : 카테고리별 상품 리스트 조회 및 베스트 상품 리스트 조회

Ko-udon opened the pull request:
## 변경사항

1. 서버 구동시 상품 데이터를 일괄처리 합니다.
2. 서버 구동시 베스트 상품Id 목록을 redis를 활용해 캐시화 하여 저장합니다.
3. 베스트 상품 업데이트는 매일 자정 2시에 이루어 집니다.


## 체크 리스트 (Checklist)

pull request 전에 아래 체크 리스트들을 만족하는 지 확인한 후 체크('x') 표시를 해주시기 바랍니다.

- [x]  master 브랜치가 아니라 **develop** 브랜치에 Merge하도록 pull request를 작성 중이신가요?
- [x]  모든 **테스트**가 성공했나요?

개발 과정에서 임시로 상품 데이터가 필요하여, 서버 구동시 상품데이터를 db에 넣습니다.
![image](https://user-images.githubusercontent.com/79897135/218314489-9e79fa66-486b-48fd-a5d3-083b42ade24e.png)

베스트 상품 목록을 매일 자정2시에 업데이트 합니다.
![image](https://user-images.githubusercontent.com/79897135/218314510-6bd68e72-6c31-4663-9cae-b04d7c3f2e87.png)

(Redis에 베스트 상품Id들이 저장되어있는 모습)
![image](https://user-images.githubusercontent.com/79897135/218313209-2c8c1e62-70ce-4479-9931-e70ea7242e20.png)


=================================================================================
카테고리별 제품 리스트 조회
(요청의 Page, Size 값에 맞게 결과값을 반환합니다)
![image](https://user-images.githubusercontent.com/79897135/218321959-40349b2f-71a5-4409-995f-cb66a76262e2.png)
![image](https://user-images.githubusercontent.com/79897135/218322024-a2c8b120-cb23-441f-b154-43b62821e183.png)


베스트 상품 리스트 조회
![image](https://user-images.githubusercontent.com/79897135/218322042-5939c21f-2396-4a9d-87c6-3580196c0681.png)
![image](https://user-images.githubusercontent.com/79897135/218322065-9275ee3e-935b-4973-b292-3a8d07d85298.png)

=================================================================================
상품 추가 api
![image](https://user-images.githubusercontent.com/79897135/218727380-29a18c93-da7f-4e8c-b8da-d497343a0b08.png)
(상품 카테고리는 저희가 정한 코드값으로 넣으면 됩니다.)
- 0 : 의류
- 1 : 가전
- 2 : 고가 레고 (완구)
- 3 : IT
- 4 : 귀금속
- 5 : 가구
- 6 : 기타

ehddn committed:
fix(product_modify): product, productLikes, review 수정

1. productLikes에 @NoArgsConstructor(access = AccessLevel.PROTECTED) 접근제한설정

2. review 의 변수명 소문자로 변경

3. review 의 @Setter 제거

4. review 에 BaseEntity 상속

ehddn committed:
feat(product): 상품 일괄처리 및 베스트 상품 업데이트

1. 서버 구동시 상품 데이터를 일괄처리 합니다.
2. 서버 구동시 베스트 상품Id 목록을 redis를 활용해 캐시화 하여 저장합니다.
3. 베스트 상품 업데이트는 매일 자정 2시에 이루어 집니다.

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/config/BatchJob.java:
```diff
@@ -0,0 +1,86 @@
+    public Step InputProductStep() {
+        return stepBuilderFactory.get("InputProductStep")
+            .tasklet((contribution, chunkContext) -> {
+                log.debug("===== 상품 데이터 저장 ====");
```
log.info로 수정 부탁드려요

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/config/BatchJob.java:
```diff
@@ -0,0 +1,86 @@
+    public Step UpdateBestProductStep() {
+        return stepBuilderFactory.get("UpdateBestProductStep")
+            .tasklet((contribution, chunkContext) -> {
+                log.debug("===== 베스트 상품 업데이트 ====");
```
마찬가지로 log.info로  수정해주세요

Ko-udon committed:
Merge branch 'dev' into PRODUCT_MODIFY_BEST&Category

ehddn committed:
feat(product): 상품 일괄처리 및 베스트 상품 업데이트

1. Batch, Scheduler에 작성된 log.dedug -> log.info 로 수정

2. 카테고리별 상품 리스트 조회(페이지 당 12개)

3. 베스트 상품 리스트 조회(12개)

ehddn committed:
Merge remote-tracking branch 'origin/PRODUCT_MODIFY_BEST&Category' into PRODUCT_MODIFY_BEST&Category

ehddn committed:
feat(product): 상품 일괄처리 및 베스트 상품 업데이트

1. Batch, Scheduler에 작성된 log.dedug -> log.info 로 수정

2. 카테고리별 상품 리스트 조회(페이지 당 12개)

3. 베스트 상품 리스트 조회(12개)

Kim-SeonBeom commented on src/main/java/com/zerobase/wishmarket/domain/product/repository/ProductRepository.java:
```diff
@@ -1,10 +1,21 @@
 @Repository
 public interface ProductRepository extends JpaRepository<Product, Long> {
 
+    Page<List<Product>> findAllByCategory(ProductCategory category, Pageable pageable);
```
Page<>라이브러리는 객체를 content에 넣을때 List에 넣어서 반환을 한다고합니다.
따라서 Page<List<Product>>는 List<List<Product>>로 return을 해주게 됩니다.
고려하시는 상황이 있는게 아니라면 Page<Product>로 바꾸시는게 좋을것 같습니다.

Ko-udon replied:
감사합니다! 수정하겠습니다~

Kim-SeonBeom commented on src/main/java/com/zerobase/wishmarket/domain/product/exception/ProductErrorCode.java:
```diff
@@ -10,6 +10,8 @@
 @RequiredArgsConstructor
 public enum ProductErrorCode implements ErrorCode {
     PRODUCT_NOT_FOUND(HttpStatus.BAD_REQUEST,"해당하는 상품이 존재하지 않습니다."),
+    BEST_PRODUCT_NOT_FOUND(HttpStatus.BAD_REQUEST, "배스트 상품을 불러오는데 실패하였습니다."),
```
오타가 있네요
베스트 상품으로 수정해주세요

Ko-udon replied:
확인하였습니다!

ehddn committed:
fix(product): 상품데이터 Best여부 추가 및 베스트 상품 업데이트 수정

1. 상품데이터에 베스트상품의 여부를 판별하기 위한 istBest변수 추가

2. 베스트 상품 업데이트시 true에 해당하는 모든
상품을 false로 초기화

3. 베스트 상품 업데이트 테스트

ehddn committed:
fix(product): PR 내용 반영

1. ProductException에 "배스트" -> "베스트" 로 오타 수정

2. 카테고리별 제품 목록 반환값을
Page<List<Product>> -> Page<Product> 로 수정

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/service/ProductService.java:
```diff
@@ -54,21 +54,54 @@ public void addProduct() {
+        setOperations = redisTemplate.opsForSet();
+
+        //베스트 상품을 삭제하기 전에, 상품 데이터에 isBest가 true인 데이터들을 false로 초기화
+        List<Product> products = productRepository.findAll();
```
jpa 에서 True 조건 걸어서 쿼리 하실 수 있을거예요

Kim-SeonBeom commented on src/main/java/com/zerobase/wishmarket/domain/product/controller/ProductController.java:
```diff
@@ -23,14 +23,14 @@ public class ProductController {
-    public Page<List<Product>> getProductListByCategory(
+    public Page<Product> getProductListByCategory(
         @RequestParam ProductCategory categories,
         @RequestParam("page") Integer page, @RequestParam("size") Integer size) {
```
size를 FE분들과 12개로 정했던거 같아서
parameter로 size를 따로 안받는게 좋을꺼같은데 어떻게 생각하시나요?

ehddn committed:
fix(product): 베스트 상품 redis에 객체로 저장

redis에서 베스트 상품 id 목록을 가지지 않고

redis에서 바로 베스트 상품 객체목록들을 가지고 조회 가능하도록
수정하였습니다.

ehddn committed:
fix(product): 상품 추가 api

백앤드 단에서 사용할 상품 넣기 api 입니다.

상품데이터를 넣을때 카테고리는 코드값으로 넣으면 됩니다.

ehddn committed:
fix(product): 베스트 상품 Batch 빠진부분 추가

베스트 상품을 업데이트 할때

각 상품 데이터의 isBest값을 Best여부에 맞게 변경하는 로직이 빠져 추가하였습니다.

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/service/ProductService.java:
```diff
@@ -28,18 +27,17 @@ public class ProductService {
     //임시로 제품 정보 넣기, Batch로 서버 시작시 데이터 저장, 추후 변경
     public void addProduct() {
         for (ProductCategory category : ProductCategory.values()) {
             for (int i = 1; i < 21; i++) {
```
반복문으로 매번 save하는 것보단 saveAll을 활용하시면 좋을거 같습니다.
